protocols/identify/src/identify.rs

```diff
@@ -153,81 +165,91 @@ where
-                protocol_version: self.protocol_version.clone(),
-                agent_version: self.agent_version.clone(),
-                listen_addrs,
-                protocols,
-            };
-
-            let future = sender.send(send_back_info, &observed);
-            self.futures.push((peer_id, future));
-        }
-
-        // Removes each future one by one, and pushes them back if they're not ready.
-        for n in (0..self.futures.len()).rev() {
-            let (peer_id, mut future) = self.futures.swap_remove(n);
-            match future.poll() {
-                Ok(Async::Ready(())) => {
-                    let event = IdentifyEvent::SendBack {
-                        peer_id,
-                        result: Ok(()),
-                    };
-                    return Async::Ready(NetworkBehaviourAction::GenerateEvent(event));
-                },
-                Ok(Async::NotReady) => self.futures.push((peer_id, future)),
-                Err(err) => {
-                    let event = IdentifyEvent::SendBack {
-                        peer_id,
-                        result: Err(err),
-                    };
-                    return Async::Ready(NetworkBehaviourAction::GenerateEvent(event));
-                },
+            let mut sending = 0;
+            let to_send = self.pending_replies.len() + 1;
+            let mut reply = Some(r);
+            loop {
+                match reply {
+                    Some(Reply::Queued { peer, io, observed }) => {
+                        let info = IdentifyInfo {
+                            public_key: self.local_public_key.clone(),
+                            protocol_version: self.protocol_version.clone(),
+                            agent_version: self.agent_version.clone(),
+                            listen_addrs: listen_addrs.clone(),
+                            protocols: protocols.clone(),
+                        };
+                        let io = io.send(info, &observed);
+                        reply = Some(Reply::Sending { peer, io });
+                    }
+                    Some(Reply::Sending { peer, mut io }) => {
+                        sending += 1;
+                        match io.poll() {
+                            Ok(Async::Ready(())) => {
+                                let event = IdentifyEvent::Sent { peer_id: peer };
+                                return Async::Ready(NetworkBehaviourAction::GenerateEvent(event));
+                            },
+                            Ok(Async::NotReady) => {
+                                self.pending_replies.push_back(Reply::Sending { peer, io });
+                                if sending == to_send {
+                                    // All remaining futures are NotReady
+                                    break
+                                } else {
+                                    reply = self.pending_replies.pop_front();
```

Comment:
This whole system where you push and pop is quite complicated to understand.
I guess it's fine, because with async/await it should be possible to rewrite it in a very straight-forward way.

Reply:
I admit I had wished for it to be simpler, but I hope it is still a small improvement over the two separate collections and loops that were there before, as it makes the state transition explicit in the `Reply` enum and a single loop, together with the use of named fields instead of tuples.